### DISCUSSION TITLE
Refactor `firstOrCreate` to include soft deleted models

### DISCUSSION THREAD

**Discussion by chrisloftus** [laravel/framework#48799] [Ideas]
I recently ran into a bug where I had a model which uses soft deletes and some simple code like:

```
$user = User::firstOrCreate([
    'email' => request()->input('user.email'),
]);

if ($user->wasRecentlyCreated) {
    info("User created: {$user->email}");
}
```

This was causing unique constraint errors at the database level when a user had previously been soft deleted as `firstOrCreate` doesn't take into account soft deletes (/`withTrashed()`).

I had to refactor the code to a much less elegant way:

```
$user = User::query()
                    ->withTrashed()
                    ->where('email', request()->input('user.email'))
                    ->first();

    if (! $user) {
                    $user = User::create(['email' => request()->input('user.email')]);

                    info("User created: {$user->email}");
                }
```

I am willing to PR this, potentially by making the following work, if people are open to it?

```
$user = User::withTrashed()->firstOrCreate([
    'email' => request()->input('user.email'),
]);
```
Thanks.

**Comment by macropay-solutions** [comment]
the solution is there already

    /**
     * Add the restore-or-create extension to the builder.
     *
     * @param  \Illuminate\Database\Eloquent\Builder  $builder
     * @return void
     */
    protected function addRestoreOrCreate(Builder $builder)
    {
        $builder->macro('restoreOrCreate', function (Builder $builder, array $attributes = [], array $values = []) {
            $builder->withTrashed();

            return tap($builder->firstOrCreate($attributes, $values), function ($instance) {
                $instance->restore();
            });
        });
    }

\Illuminate\Database\Eloquent\SoftDeletingScope::addRestoreOrCreate

**Comment by shaedrich** [comment]
The other option would be to make the field semi unique:

```php
<?php
 
use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;
 
return new class extends Migration
{
    /**
     * Run the migrations.
     */
    public function up(): void
    {
        Schema::create('users', function (Blueprint $table) {
            $table->id();
            $table->string('email')->check();
            $table->boolean('ucode')->virtualAs("CASE WHEN deleted_at IS NULL THEN b'1' ELSE NULL END")
            $table->unique([ 'ucode', 'email' ]);
            $table->softDeletes();
            $table->timestamps();
        });
    }
};
```

<small>(see https://dba.stackexchange.com/questions/167544/how-to-create-a-unique-index-with-only-certain-values-required-to-be-unique)</small>
